### PR TITLE
Test and fix for moving non-euclidean axes in arcs

### PR DESCRIPTION
**Pull request by lucalenardi**
We are using beagleg (thanks for this super nice project, btw!) to drive stepper motors systems for industrial machines. In most cases, the system is made by XYZ axis, plus one or more axes to drive single or multiple extruders (A, B, E etc). Those tools needs to be synced with the motion during the process, that's why interpolating A, B... axes with X, Y, Z is actually needed.
 
It works perfectly for points and lines, but not with arcs, where non-euclidean axes are ignored.

This pull request tries to solve this issue by adding a test (with an arc on the XY plane and a linear motion for A) and a small fix in `arc_gen.cc` to make it pass.

### CODE REVIEW

File: src/gcode-parser/arc-gen_test.cc [hunk 1]
```diff
@@ -154,6 +155,28 @@ static void testArcLength(GCodeParserAxis normal, bool clockwise, double start,
   }
 }
 
+TEST(ArcGenerator, ArcLength_NonEuclideanAxes) {
```
Comment (hzeller) [inline src/gcode-parser/arc-gen_test.cc:158]: To make more clear what we're testing, I'd probably name it
`ArcLength_InterpolateNonEuclidianAxes`.

File: src/gcode-parser/arc-gen_test.cc [hunk 2]
```diff
@@ -154,6 +155,28 @@ static void testArcLength(GCodeParserAxis normal, bool clockwise, double start,
...
+  AxesRegister start, center, target;
+  const double kTurnAngle = M_PI;
+  const float kAxisALenght = 20.0;
```
Comment (hzeller) [inline src/gcode-parser/arc-gen_test.cc:161]: Lenght : typo, last two letters swapped.

File: src/gcode-parser/arc-gen_test.cc [hunk 3]
```diff
@@ -59,6 +59,7 @@ class TestArcAccumulator : public GCodeParser::EventReceiver {
   }
 
   float total_len() const { return total_len_; }
+  AxesRegister last() const { return last_; }
```
Comment (hzeller) [inline src/gcode-parser/arc-gen_test.cc:62]: I'd make that a `const AxesRegister&`  reference.

**Comment by hzeller** [comment]
How do you define the additional axis in the `G2` or `G3` call ?

I guess you can just give it an `A` position as target position, but I think that is never tested anywhere.
So I suggest to also add a test for that in [gcode-parser_test.cc](https://github.com/hzeller/beagleg/blob/4693e81c0271b76292bcb8455824e01ee0f7ba50/src/gcode-parser/gcode-parser_test.cc#L600) so that this feature is documented.

**Comment by hzeller** [comment]
I suppose the axes other than X, Y and Z would also be good to interpolate for the `G5` and `G5.1` spline features. I am pretty sure it has the same problem that you fixed for the circles. If you want to have a follow-up pull request :)

**Comment by hzeller** [comment]
to fix the formatting problem the continuous integration complains about, just run

```
.github/bin/run-clang-format.sh
```

**Comment by hzeller** [comment]
Alright, I checked: the gcode parser was also not taking the additional axis into account. That is now fixed in head ([this was the change](https://github.com/hzeller/beagleg/commit/b377e316473f9c5fb4014ba0532e484886fc4965) ... including some unrelated clean-up and improved tests).

I have added a [section in the test](https://github.com/hzeller/beagleg/blob/bec5bf057b63bc553acdbf237c704215a458889d/src/gcode-parser/gcode-parser_test.cc#L700-L704) that now fails (so this is why it is `#if`-def'ed out), but should pass after your change is in.

**Comment by lucalenardi** [comment]
> I suppose the axes other than X, Y and Z would also be good to interpolate for the `G5` and `G5.1` spline features. I am pretty sure it has the same problem that you fixed for the circles. If you want to have a follow-up pull request :)

Yes definitely something I will look at.

**Review by hzeller — APPROVED** [review]
LGTM

**Comment by hzeller** [comment]
Thanks! Merged.

You now have the honor of removing the `#if 0` around here, as it will now be working
https://github.com/hzeller/beagleg/blob/1cf62595bf927bc221f991adc547cbe8eeac329f/src/gcode-parser/gcode-parser_test.cc#L700-L704

**Pull request merged**